specs/language/basic.tex

```diff
@@ -1,5 +1,97 @@
 \Ch{Basic Concepts}{Basic}
 
+\Sec{Types}{Basic.types}
+
+\p The \textit{object representation} of an object of type \texttt{T} is the
+sequence of \textit{N} bytes taken up by the object of type \texttt{T}, where
+\textit{N} equals \texttt{sizeof(T)}\footnote{\texttt{sizeof(T)}, returns the
+size of an object of type \texttt{T} in bytes as stored in device memory.
+The size may be different in other memory spaces and no operator exists to query
+it.}. The \textit{object representation} of an object may be different based on
+the \textit{memory space} it is stored in (\ref{Intro.Memory.Spaces}).
+
+\p The \textit{value representation} of an object is the set of bits that hold
+the value of type \texttt{T}.
+
+\p An \textit{object type} \texttt{T} is an \textit{incomplete type} if the
+compiler lacks sufficient information to determine the size of an object of type
+\texttt{T}, and a \textit{complete type} if the compiler has sufficient
+information to determine the size of an object of type \texttt{T}. An object may
+not be defined to have an \textit{incomplete} type.
+
+\p A \textit{class type} is a data type declared with either the \texttt{class}
+or \texttt{struct} keywords (\ref{Classes}). A class type \texttt{T} may be
+declared as incomplete at one point in a translation unit via a \textit{forward
+declaration}, and complete later with a full definition. The type \texttt{T} is
+the same type throughout the translation unit.
+
+\p An \textit{object type} is a type that is not a function type, not a
+reference type, and not a void type.
+
+\p Arithmetic types (\ref{Basic.types.arithmetic}), enumeration types, and
+\textit{cv-qualified} versions of these types are collectively called
+\textit{scalar types}.
+
+\p Vectors of scalar types declared with the built-in \texttt{vector<T,N>}
+template are \textit{vector types}.
+
+\p \textit{POD types}\footnote{POD is an initialism for Plain-Old-Data which is
+referenced in the C++ specification but never expanded.} are POD class types,
+scalar types, vector types, and arrays of such types. \textit{POD class types}
+are class types that only contain members POD types.
```

Comment:
C++ moved away from the "POD" nomenclature in C++20 (they split it into "trivial" and "standard-layout" types - the old "POD" types are types that are both trivial and standard-layout). We should probably either follow their lead or not bother with this definition at all. I suspect we can get away with leaving this out, at least for now.

Reply:
I struggled with this a bunch. Part of the reason I went with POD wording was because I felt it was clear to understand and the current C++ wording gets _way_ complicated because of resource types.

Technically resource types are trivially copyable, types containing resources are trivially copyable, but they also have magic special behavior and can't be used in some contexts.

It makes sense to me that C++ focuses on the ability to copy values of the types, but that doesn't quite give the distinction we need for HLSL.